File: src/java/VRE.Vridge.API/VRE.Vridge.API.Client/src/main/java/com/riftcat/vridge/api/client/java/remotes/ControllerRemote.java
package com.riftcat.vridge.api.client.java.remotes;

import com.riftcat.vridge.api.client.java.proto.HandType;
import com.riftcat.vridge.api.client.java.proto.HeadRelation;
import com.riftcat.vridge.api.client.java.proto.VRController;
import com.riftcat.vridge.api.client.java.proto.VRControllerAxis_t;
import com.riftcat.vridge.api.client.java.proto.VRControllerState_t;
import com.riftcat.vridge.api.client.java.proto.VRControllerState_tOrBuilder;
import com.riftcat.vridge.api.client.java.proxy.ControllerProxy;
import com.riftcat.vridge.api.client.java.utils.ButtonMask;

import java.util.concurrent.TimeoutException;

public class ControllerRemote extends RemoteBase {

    private static int packetNum = 0;
    private ControllerProxy proxy;

    ControllerRemote(ControllerProxy proxy) {
        super(proxy);
        this.proxy = proxy;
    }

    /**
     * Sets VR controller to a new state
     * @param controllerId Unique ID of given controller
     * @param headRelation How given pose relates to the head. Usually unrelated is the best pick.
     * @param orientation Orientation as XYZW float[4].
     * @param position Position as XYZ float[3].
     * @param analogX (-1,1) horizontal touchpad position.
     * @param analogY (-1,1) vertical touchpad position.
     * @param analogTrigger (0,1) trigger state (1 is fully pulled)
     * @param isMenuPressed
     * @param isSystemPressed
     * @param isTriggerPressed
     * @param isGripPressed
     * @param isTouchpadPressed
     * @param isTouchpadTouched
     */
    public void setControllerState(
            // Controller ID
            int controllerId,
            HandType handType,
            boolean disableController,
            // Pose data
            HeadRelation headRelation,
            float[] orientation,
            float[] position,
            // Touchpad state [-1,1]
            double analogX,
            double analogY,

            // Trigger state
            double analogTrigger,

            // Button states
            boolean isMenuPressed,
            boolean isSystemPressed,
            boolean isTriggerPressed,
            boolean isGripPressed,
            boolean isTouchpadPressed,
            boolean isTouchpadTouched){

        // See openvr.h in OpenVR SDK for mappings and masks
        // https://github.com/ValveSoftware/openvr/blob/master/headers/openvr.h

        long pressedMask = buildPressedMask(isMenuPressed, isSystemPressed, isTriggerPressed, isGripPressed, isTouchpadPressed);
        long touchedMask = buildTouchedMask(isTouchpadTouched, isTriggerPressed);

        int controllerStatus = 0;
        if(disableController == true) {
            controllerStatus = 2;
        }

        VRControllerState_tOrBuilder buttons = VRControllerState_t.newBuilder()
                .setRAxis0(VRControllerAxis_t.newBuilder()
                        .setX((float) analogX)
                        .setY((float) analogY))
                .setRAxis1(VRControllerAxis_t.newBuilder()
                        .setX((float) analogTrigger))
                .setUlButtonPressed(pressedMask)
                .setUlButtonTouched(touchedMask)
                .setUnPacketNum(++packetNum);      // Touchpad


        VRController.Builder controllerState = VRController.newBuilder()
                .setControllerId(controllerId)
                .addOrientation(orientation[0])
                .addOrientation(orientation[1])
                .addOrientation(orientation[2])
                .addOrientation(orientation[3])
                .setStatus(controllerStatus)
                .setSuggestedHand(handType)
                .setHeadRelation(headRelation)
                .setButtonState((VRControllerState_t.Builder) buttons);

        if(position != null){
            controllerState.addPosition(position[0]);
            controllerState.addPosition(position[1]);
            controllerState.addPosition(position[2]);
        }

        try {
            proxy.sendControllerState(controllerState.build());
        } catch (Exception e) {
            dispose();
        }
    }

    /** Recenter head tracking. Works the same as pressing recenter hotkey as configured in VRidge settings. */
    public void recenterHead(){
        try{
            proxy.recenterHead();
        }
        catch (Exception e){
            dispose();
        }
    }

    private long buildTouchedMask(boolean isTouchpadTouched, boolean isTriggerTouched) {
        long mask = 0;
        if (isTouchpadTouched) mask |= ButtonMask.Touchpad;
        if (isTriggerTouched)  mask |= ButtonMask.Trigger;
        return mask;
    }

    private long buildPressedMask(
            boolean isMenuPressed,
            boolean isSystemPressed,
            boolean isTriggerPressed,
            boolean isGripPressed,
            boolean isTouchpadPressed) {

        long mask = 0;
        if (isMenuPressed) mask |= ButtonMask.ApplicationMenu;
        if (isSystemPressed) mask |= ButtonMask.System;
        if (isTriggerPressed) mask |= ButtonMask.Trigger;
        if (isGripPressed) mask |= ButtonMask.Grip;
        if (isTouchpadPressed) mask |= ButtonMask.Touchpad;
        return mask;
    }
}

File: src/csharp/VRE.Vridge.API.Client/Remotes/ControllerRemote.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Numerics;
using System.Text;
using VRE.Vridge.API.Client.Helpers;
using VRE.Vridge.API.Client.Messages.BasicTypes;
using VRE.Vridge.API.Client.Messages.OpenVR;
using VRE.Vridge.API.Client.Messages.v3.Broadcast;
using VRE.Vridge.API.Client.Messages.v3.Controller;
using VRE.Vridge.API.Client.Proxy.Controller;
using VRController = VRE.Vridge.API.Client.Messages.v3.Controller.VRController;

namespace VRE.Vridge.API.Client.Remotes
{
    public class ControllerRemote : RemoteBase<ControllerProxy>
    {
        private uint packetNum = 0;                    

        internal ControllerRemote(ControllerProxy proxy) : base(proxy)
        {
        }

        /// <summary>
        /// Sends full controller state as defined by arguments.
        /// </summary>        
        public void SetControllerState(
            // Controller ID
            int controllerId,

            // Pose data
            HeadRelation headRelation,
            HandType suggestedHand,
            Quaternion orientation,
            Vector3? position,

            // Touchpad state [-1,1]
            double analogX,
            double analogY,

            // Trigger state
            double analogTrigger,

            // Button states
            bool isMenuPressed,
            bool isSystemPressed,
            bool isTriggerPressed,
            bool isGripPressed,
            bool isTouchpadPressed,
            bool isTouchpadTouched)
        {

            // See openvr.h in OpenVR SDK for mappings and masks
            // https://github.com/ValveSoftware/openvr/blob/master/headers/openvr.h

            var buttons = new VRControllerState_t()
            {
                rAxis0 = new VRControllerAxis_t((float)analogX, (float)analogY),       // Touchpad
                rAxis1 = new VRControllerAxis_t((float)analogTrigger, 0),              // Trigger
                rAxis2 = new VRControllerAxis_t(0, 0),
                rAxis3 = new VRControllerAxis_t(0, 0),
                rAxis4 = new VRControllerAxis_t(0, 0),
                ulButtonPressed = BuildButtonPressedMask(isMenuPressed, isSystemPressed, isTriggerPressed, isGripPressed, isTouchpadPressed),
                ulButtonTouched = BuildButtonTouchedMask(isTouchpadTouched, true),
                unPacketNum = ++packetNum

            };

            var controllerState = new VRController()
            {
                ButtonState = buttons,
                Status = 0,
                ControllerId = controllerId,
                Position = position.HasValue ? new[] { position.Value.X, position.Value.Y, position.Value.Z } : null,
                Orientation = new [] {orientation.X, orientation.Y, orientation.Z, orientation.W },
                HeadRelation = headRelation,
                SuggestedHand = suggestedHand
            };

            WrapTimeouts(() => { Proxy.SendControllerData(controllerState); });
        }

        /// <summary>
        /// Recenter head tracking. Works the same as pressing recenter hotkey as configured in VRidge settings.
        /// </summary>
        public void RecenterHead()
        {

            WrapTimeouts(() => { Proxy?.RecenterHead(); });

        }

        internal override void Dispose()
        {
            Proxy?.Disconnect();
            base.Dispose();
        }

        /// <summary>
        /// Maps button pressed state into OpenVR packed ulong.
        /// </summary>                
        private ulong BuildButtonPressedMask(
            bool isMenuPressed,
            bool isSystemPressed,
            bool isTriggerPressed,
            bool isGripPressed,
            bool isTouchpadPressed)
        {
            ulong mask = 0;            

            if (isMenuPressed) mask |= ButtonMask.ApplicationMenu;
            if (isSystemPressed) mask |= ButtonMask.System;
            if (isTriggerPressed) mask |= ButtonMask.Trigger;
            if (isGripPressed) mask |= ButtonMask.Grip;
            if (isTouchpadPressed) mask |= ButtonMask.Touchpad;

            return mask;
        }


        /// <summary>
        /// Maps button touched state into OpenVR packed ulong.
        /// </summary>         
        private ulong BuildButtonTouchedMask(bool isTouchpadTouched, bool isTriggerTouched)
        {
            ulong mask = 0;

            if (isTouchpadTouched) mask |= ButtonMask.Touchpad;
            if (isTriggerTouched) mask |= ButtonMask.Trigger;

            return mask;
        }

        /// <summary>
        /// Builds 4x4 combined matrix for given translation and rotation.
        /// Not used and not recommended anymore. Send orientation[+position] instead of matrix.
        /// </summary>        
        private Matrix4x4 BuildControllerMatrix(double x, double y, double z, double yaw, double pitch, double roll)
        {
            Matrix4x4 m = Matrix4x4.Identity;

            
            m *= Matrix4x4.CreateFromYawPitchRoll(
                (float)MathHelpers.DegToRad(yaw),
                (float)MathHelpers.DegToRad(pitch),
                (float)MathHelpers.DegToRad(roll));

            m*= Matrix4x4.CreateTranslation((float)x, (float)y, (float)z);            

            return m;
        }
    }
}
